### PR TITLE
Alias support for Document Level Monitors

### PR DESCRIPTION
**Pull request by lezzago**
*Issue #, if available:*
N/A
*Description of changes:*
- Implemented support for defining doc level monitors using aliases.

*CheckList:*
[X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt [hunk 1]
```diff
@@ -69,47 +71,82 @@ class DocLevelMonitorQueries(private val client: AdminClient, private val cluste
...
+                        if (clusterState.routingTable.hasIndex(indexName)) {
+                            val indexMetadata = clusterState.metadata.index(indexName)
+
+                            if (indexMetadata.mapping() != null) {
+                                log.info("Index name: $indexName")
+                                val properties = (
+                                    (indexMetadata.mapping()?.sourceAsMap?.get("properties"))
+                                        as Map<String, Map<String, Any>>
+                                    )
+                                val updatedProperties = properties.entries.associate { "${it.key}_$monitorId" to it.value }.toMutableMap()
```
Comment (sbcd90) [inline alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt:95]: this will result in a conflict if two or more indices have the same property name. 
the conflict can be resolved by adding index keys to the property names.
instead of `${it.key}_$monitorId`, we can use this `${it.key}_${indexName}_$monitorId`

Comment (lezzago) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt]: done

**Review by sbcd90 — APPROVED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt [hunk 2]
```diff
@@ -69,47 +71,82 @@ class DocLevelMonitorQueries(private val client: AdminClient, private val cluste
...
+                                    updateMappingRequest,
+                                    object : ActionListener<AcknowledgedResponse> {
+                                        override fun onResponse(response: AcknowledgedResponse) {
+
+                                            queries.forEach {
+                                                var query = it.query
+
+                                                properties.forEach { prop ->
+                                                    query = query.replace("${prop.key}:", "${prop.key}_$monitorId:")
+                                                }
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt:110]: Should query property now also have the index name now? Given, now queries in the query document will be repeated per index, do we need an additional filtering along with the monitor_id?

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentReturningMonitorRunner.kt [hunk 3]
```diff
@@ -273,6 +294,21 @@ object DocumentReturningMonitorRunner : MonitorRunner() {
         return finding.id
     }
 
+    private fun updateLastRunContext(
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/DocumentReturningMonitorRunner.kt:297]: will this require any change in the monitor schema, since we have per index context to be persisted?

File: alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt (hunk 2 above)
Comment (lezzago) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt:110]: Yes that has been added from Subho's comment. I think you are not seeing the latest commit

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentReturningMonitorRunner.kt (hunk 3 above)
Comment (lezzago) [reply alerting/src/main/kotlin/org/opensearch/alerting/DocumentReturningMonitorRunner.kt:297]: This does not require further changes to the monitor schema

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentReturningMonitorRunner.kt [hunk 4]
```diff
@@ -83,62 +83,86 @@ object DocumentReturningMonitorRunner : MonitorRunner() {
...
+                val indexLastRunContext = lastRunContext.getOrPut(indexName) {
+                    createRunContext(monitorCtx.clusterService!!, monitorCtx.client!!, indexName)
+                }
+
+                // Prepare updatedLastRunContext for each index
+                val indexUpdatedRunContext = updateLastRunContext(
+                    indexLastRunContext.toMutableMap(),
+                    monitorCtx,
+                    indexName
+                ) as MutableMap<String, Any>
```
[the remark this answers is not in the document]
Comment (lezzago) [reply alerting/src/main/kotlin/org/opensearch/alerting/DocumentReturningMonitorRunner.kt:117]: Yes, that is the current expected failure scenario. We can still discuss this and change it so that it will still execute for the successful indices.

[the remark this answers is not in the document]
Comment (getsaurabh02) [reply alerting/src/main/kotlin/org/opensearch/alerting/DocumentReturningMonitorRunner.kt:117]: Will highly recommend that. Lets add this out our quick follow-up list.

File: alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt (hunk 2 above)
Comment (getsaurabh02) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt]: Just saw the update, thanks. I hope we are considering this during index creation, update time as well.

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetFindingsAction.kt [hunk 5]
```diff
@@ -90,12 +91,23 @@ class TransportGetFindingsSearchAction @Inject constructor(
...
+                .should(
+                    QueryBuilders.nestedQuery(
+                        "queries",
+                        QueryBuilders.boolQuery()
+                            .must(
+                                QueryBuilders
+                                    .queryStringQuery(tableProp.searchString)
+                                    .defaultOperator(Operator.AND)
+                                    .field("queries.tags")
+                                    .field("queries.name")
```
Comment (AWSHurneyt) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetFindingsAction.kt:107]: Were we still planning to implement support for searching for findings using a specific `monitor_id`, or would that be better suited for a separate PR?

Comment (lezzago) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetFindingsAction.kt:107]: The part in line 96 supports `monitor_id`, I added a integ test to showcase that

File: alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt [hunk 6]
```diff
@@ -69,47 +71,83 @@ class DocLevelMonitorQueries(private val client: AdminClient, private val cluste
...
-                                    .source(mapOf("query" to mapOf("query_string" to mapOf("query" to query)), "monitor_id" to monitorId))
-                                request.add(indexRequest)
+        val getAliasesRequest = GetAliasesRequest(index)
+        queryClient.admin().indices().getAliases(
+            getAliasesRequest,
+            object : ActionListener<GetAliasesResponse> {
+                override fun onResponse(getAliasesResponse: GetAliasesResponse?) {
+                    val aliasIndices = getAliasesResponse?.aliases?.keys()?.map { it.value }
+                    val isAlias = aliasIndices != null && aliasIndices.isNotEmpty()
+                    val indices = if (isAlias) getAliasesResponse?.aliases?.keys()?.map { it.value } else listOf(index)
```
Comment (AWSHurneyt) [inline alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt:82]: Minor question; would it make sense to refactor this to the following since the response has already been parsed to create `aliasIndices`?
```
val indices = if (isAlias) aliasIndices else listOf(index)
```

**Review by AWSHurneyt — APPROVED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt (hunk 6 above)
Comment (lezzago) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/DocLevelMonitorQueries.kt]: Nice catch
